RFC-0008-plan-checker.md

```diff
@@ -0,0 +1,171 @@
+* An optional string representing the serialized converted plan fragment.  Presence indicates that the plan was successfully
+  converted to a Velox plan fragment, absence indicates that the plan was not converted.
+
+### Presto to Velox plan validation
+
+The Presto runtime centralizes plan validation logic into the `PlanChecker` class.  There exist three phases to this class: 
+
+* `validateIntermediatePlan`
+* `validateFinalPlan`
+* `validateFragmentedPlan`
+
+It is generally useful to allow this class to be configured with an SPI that allows for custom plan validation logic to be added.
+For example, a business may decide that a certain type is not allowed, or add a check to ensure that plans that are overly
+complicated are killed.
+
+An SPI will be added that will add more checks to the `PlanChecker` class that will allow additional checks for each of the planning
+phases.  The SPI will contain a field indicating which phase of the plan checker to be added, and a validator that will be run during
+that phase.
+
+A new endpoint to the Presto sidecar will be added that will attempt to convert a Presto plan to a Velox plan.  In the
+`presto-native-plugin` module, a new implementation of the SPI will be added which will add a check to the `PlanChecker` class
+which will call the sidecar to attempt to convert the plan.  If the conversion fails, the plan checker will fail the plan.  It will
+be added at the `validateFragmentedPlan` phase--this is because it's not until plan fragmentation occurs that we know which portions
+of the plan will be executed in the coordinator, and which will be executed in the workers.  Plan fragments that are executed in the
+coordinator, such as `COORDINATORY_ONLY` distribution types, will be skipped by the plugin.
+
+The code for the plan checker will run `PrestoToVeloxQueryPlan`, which is used by the workers to convert the Presto plan fragment
+to a Velox plan fragment.  If the conversion fails, the plan checker will fail the plan, returning with it the reason for the failure.
+
+#### Failing the plan quickly
+
+An additional code change will be made to allow the planner to execute prior to queuing.  This is so that the plan checker can
+be run before the query is queued.  This will allow the user to get feedback on the plan before the query is executed, and will
+allow the query to fail quickly if the plan is invalid.
+
+Because the queue limits concurrency, and too much concurrency during planning may require excessive resources in the coordinator,
+the plan checker will be run in a separate thread pool.  This thread pool will be configured with a maximum number of threads
+that can be run concurrently.  If the thread pool is full, then planning will wait until there is a free thread to run the planner.
+This will be configured with a new configuration parameter and session property.
+
+### EXPLAIN (TYPE DISTRIBUTED, FORMAT NATIVE)
+
```

Comment:
format should specify how the data is represented (like text or json). Type should tell you what kind of data you'll get.  We should instead add a new type `EXPLAIN(TYPE NATIVE)` that will produce the native plan (which also happens to be distributed by definition). We might provide the option of outputting the Velox plan in text or JSON format, or even graphviz if someone feels like implementing it.

Reply:
I thought about this, but decided against it:

1) Because the Velox plans are generated per fragment, it's really the fragmented plan that we are generating, which is what the distributed type gives you.
2) `FORMAT NATIVE` is meant to signify, just delegate the formatting to the underlying plugin, which in this case is just delegating to the plan converter.

Originally, I was unsure how to handle the `FORMAT` argument if specified.  But now that I think about it, perhaps what we can do is:

1) Add a new type, `DISTRIBUTED_NATIVE` or some such.  It's verbose, but it signifies that this is the distributed plan that will actually show the underlying eval's plan fragment, not Presto's version of it.
2) Add a format argument to the SPI so that we can configure the plan converter to generate different formats like Graphviz and JSON.

What do you think?

Reply:
first want to note that we aren't just delegating the formatting to the underlying plugin. We are generating a new velox plan based on each plan fragment and returning the representation of that new plan.

For your suggestions:
For 1. I prefer  `NATIVE` to `DISTRIBUTED_NATIVE` as the new type because it's less verbose, and the native plan will always be distributed because it is based on plan fragments. However, either is fine.
For 2. sounds good to me

Reply:
I updated the proposal to `EXPLAIN (TYPE NATIVE)`, and added details on how the format string would work, PTAL!